044-kraft-liveness-readiness.md

```diff
@@ -0,0 +1,198 @@
+If it has a broker process it should be listening on the replication port specified as the `inter.broker.listener.name` which is hard-coded to 9091.
+If the node is running as a combined mode, it should be listening on both ports
+
+### BrokerState
+
+The expected [BrokerState](https://github.com/apache/kafka/blob/trunk/metadata/src/main/java/org/apache/kafka/metadata/BrokerState.java) transitions in the "happy path" during a normal broker lifecycle are:
+
+NOT_RUNNING (0) -> STARTING (1) -> RECOVERY (2) -> RUNNING (3) -> PENDING_CONTROLLED_SHUTDOWN (6) -> SHUTTING_DOWN (7)
+
+The other possible value is 127 which represents UNKNOWN.
+
+You can see how the state transitions in the [BrokerLifecycleManager](https://github.com/apache/kafka/blob/trunk/core/src/main/scala/kafka/server/BrokerLifecycleManager.scala#L379), specifically that 
+the broker transitions from STARTING to RECOVERY once it has caught up with the cluster metadata, then it transitions to RUNNING as long as it is not fenced.
+
+The broker nodes cannot catch up with the cluster metadata until the controller quorum has been formed, so they will not move out of the STARTING state until the quorum leader election has happened.
+
+## Proposal
+
+There are three different types of liveness/readiness probe that we need to consider, first for a controller only, then 
+for a broker only, and finally for a combined node that has both roles.
+
+The following statements describe the intent of the proposed probes:
+
+* A controller is:
+    * "alive" if it has a process running.
+    * "ready" when it is ready to join the controller quorum.
+* A broker is:
+    * "alive" if it has a process running.
+    * "ready" when it is ready to start accepting producer/consumer requests.
+* A combined node is:
+    * "alive" if it has a process running.
+    * "ready" when it is ready to start accepting producer/consumer requests.
+        * This implies that the controller quorum will have been formed while the nodes are not marked as ready, since it cannot start accepting producer/consumer requests until it can talk to the leader.
+        * This relies on the node still accepting incoming connections from other controller nodes even if it isn't actually marked as "ready" yet.
+
+### Controller only mode
+
+The proposed probes are:
+
+* Liveness: the Java process in the container is running
+* Readiness: controller is listening on the port 9090 which is configured in `controller.listener.names`
+
+### Broker only mode
+
+The proposed probes are:
+
+* Liveness: broker is listening on the port 9091 which is configured in `inter.broker.listener.name`
+* Readiness: the BrokerState metric >=3 && != 127
+
+**Note:** This means the brokers will not become ready until a majority of the controllers 
+are up and running.
+This is similar to the current behaviour of brokers when ZooKeeper is not ready.
```

Comment:
Quick question ... With ZooKeeper it works today like this:
* When the broker is starting and ZooKeeper is not available (not running, not having a quorum), the broker crash-loops.
* When the broker was already running and ready (connected to ZooKeeper) and suddenly ZooKeeper loops quorum, Kafka will keep trying to reconnect, but it does not crash-loop or become not ready. It is still ready and does what it can without ZooKeeper.

How does this behave in KRaft - mainly in the second point? What will the readiness probe do here when it was running and ready and then the controller quorum is lost? Will it be the same behavior? Will it move into un-ready?

Reply:
From reading the KIP my understanding is that if the broker cannot contact the lead controller it will transition back into the RECOVERY state, which means yes it would move to un-ready. I think that is acceptable though, since the broker would also stop accepting client requests at that point. Whereas in ZooKeeper mode the broker could still accept client requests if it could not contact ZooKeeper. I'll do some tests this afternoon to confirm my understanding and make sure the broker state does transition back as I described.

This is the part of the KIP that I think is relevant:

> The Broker State Machine
Currently, brokers register themselves with ZooKeeper right after they start up.  This registration accomplishes two things: it lets the broker know whether it has been elected as the controller, and it lets other nodes know how to contact it.
>
> In the post-ZooKeeper world, brokers will register themselves with the controller quorum, rather than with ZooKeeper.
>
> Currently, if a broker loses its ZooKeeper session, the controller removes it from the cluster metadata.  In the post-ZooKeeper world, the active controller removes a broker from the cluster metadata if it has not sent a MetadataFetch heartbeat in a long enough time.
>
> In the current world, a broker which can contact ZooKeeper but which is partitioned from the controller will continue serving user requests, but will not receive any metadata updates.  This can lead to some confusing and difficult situations.  For example, a producer using acks=1 might continue to produce to a leader that actually was not the leader any more, but which failed to receive the controller's LeaderAndIsrRequest moving the leadership.
>
> In the post-ZK world, cluster membership is integrated with metadata updates.  Brokers cannot continue to be members of the cluster if they cannot receive metadata updates.  While it is still possible for a broker to be partitioned from a particular client, the broker will be removed from the cluster if it is partitioned from the controller.

Reply:
I've done some tests but haven't actually managed to see the BrokerState metric transition out of 3 once it got to 3. I tried:
 - A single controller and removing the controller
 - 3 controllers and taking down a controller to force an election
 - 3 controllers and taking down two controllers
 - 3 controllers and taking down all controllers

In all cases the BrokerState metric stayed at value 3

Reply:
But did you check if you can connect with clients or not? If it indeed blocks the clients, then we probably need to transition it and it should be exposed somehow through Kafka. If the clients can keep using it with some limitations, we can probably stick with being ready as we do today.

Reply:
I did some more tests this morning and it seems like the clients don't get blocked, but some requests do fail. When all controllers were down any existing producer and consumer clients were able to continue producing and consuming. However any requests that included a metadata request failed, so I couldn't start a new consumer or create a new topic. However I could create a new producer client.

Reply:
Ok, that sounds similar to ZK node. So I think we can keep it ready as we do today. Thanks for the clarification.